exploration/0003-formatted-parts.md

```diff
@@ -0,0 +1,212 @@
+[November 2022](https://bugs.chromium.org/p/v8/issues/detail?id=13494).
+
+## Use-Cases
+
+- Markup elements
+- Non-string values
+- Message post-processors
+- Decoration of placeholder interior parts.
+  For example, identifying the separate fields in these two currency values
+  (notice that the symbol, number, and fraction fields
+  are not in the same order and that the separator has been omitted):
+  ![image](https://github.com/unicode-org/message-format-wg/assets/69082/cb68c87f-9c0c-4bc6-b9a0-b1f97b2b789a)
+  ![image](https://github.com/unicode-org/message-format-wg/assets/69082/aedd4e66-7d47-4026-8b93-4ba061bb4d84)
+- Supplying bidirectional isolation of placeholders,
+  such as by using HTML's `span` element with a `dir` attribute
+  based on the direction of the placeholder.
+
+## Requirements
+
+- Define an iterable sequence of formatted part objects.
+- Include metadata for each part, such as type, source, direction, and locale.
+- Allow the representation of non-string values.
+- Allow the representation of values that consist of an iterable sequence of formatted parts.
+- Be able to represent each resolved value of a pattern with any number of formatted parts, including none.
+- Define the formatted parts in a manner that allows synonymous but appropriate implementations in different programming languages.
+
+## Constraints
+
+- The JS Intl formatters already include formatted-parts representations for each supported data type.
+  The JS implementation of the MF2 formatted-parts representation should be able to match their structure,
+  at least as far as that's possible and appropriate.
+
+## Proposed Design
+
+The formatted-parts API is included in the spec as an optional but recommended formatting target.
+
+The shape of the formatted-parts output is defined in a manner similar to the data model,
+which includes TypeScript, JSON Schema, and XML DTD definitions of the same data structure.
+
+At the top level, the formatted-parts result is an iterable sequence of parts.
+Parts corresponding to each _text_ can be simpler than those of _expressions_,
+as they do not have a `source` other than their `value`,
+or set any of the other possible metadata fields.
+
+```ts
+type MessageParts = Iterable<
+  MessageTextPart | MessageExpressionPart | MessageBiDiIsolationPart
+>;
+
+interface MessageTextPart {
+  type: "text";
+  value: string;
+}
+```
+
+For MessageExpressionPart, the `source` corresponds to the expression's fallback value.
+The `dir` and `locale` attributes of a part may be inherited from the message
+or from the operand (if present),
+or overridden by an expression attribute or formatting function,
+or otherwise set by the implementation.
+Each part should have at most one of `value` or `parts` defined;
+some may have none.
+
+```ts
+interface MessageExpressionPart {
+  type: string;
+  source: string;
+  parts?: Iterable<{ type: string; value: unknown }>;
+  value?: unknown;
+  dir?: "ltr" | "rtl" | "auto";
```

Comment:
It is unclear to me why have `dir` and `locale` at this level.

We don't need a locale to format anything, because the parts should be already formatted.
The whole proposal is called "Formatted Parts"

The locale might be needed to render things.

Or to process the formatting result (fix grammatical agreements as a post-step, fix "a apple" to "an apple" (en) or "La abeille" => "L'abbeile" (fr), or to sentence case the result of "{item} is foo"

But then that is something that is needed for the whole collection of parts, not on `MessageExpressionPart` only.

Reply:
These are needed to allow embedding content in a message that uses a different script or locale than the surrounding message.

Reply:
Kind of clear what they can be used for.

But this info is on the `MessageExpressionPart`, which comes from an expression.
And an expression can't create this info out of nothing, it is probably something we passed as a parameter.
So if I already know the info (because I passed it to the expression), having it on the `MessageExpressionPart` is useless duplication.

Reply:
Consider this message, intended for consumption by a text-to-speech system:
```
In French, the number {98 :number} is commonly expressed as {98 :number @locale=fr},
but in Belgium it's {98 :number @locale=fr-BE}.
```

How would you propose that the locale information is transmitted, if not as a field on the formatted parts?

```js
[
  { type: 'text', value: 'In French, the number ' },
  { type: 'number', source: '|98|', parts: [{ type: 'integer', value: '98' }] },
  { type: 'text', value: ' is commonly expressed as ' },
  { type: 'number', source: '|98|', locale: 'fr', parts: [{ type: 'integer', value: '98' }] },
  { type: 'text', value: ", but in Belgium it's " },
  { type: 'number', source: '|98|', locale: 'fr-BE', parts: [{ type: 'integer', value: '98' }] },
  { type: 'text', value: '.' }
]
```

As context, the `fr` number would be "quatre-vingt-dix-huit", while in `fr-BE` it's "nonante-huit".

Reply:
Language and direction are needed for placeholders because they represent values being inserted into the overall string. The language (locale) is used to ensure proper rendering and processing (such as line breaking, text transforms, or spell-checking). The direction is used to enable bidi isolation and get the direction of the substring correct.

The language and direction of a formatted part might not match that of the message due to resource fallback when looking up the message. Or because values passed in have different language or direction. (And we want bidi isolation even if the directions match!!!!)

Providing the fields in the formatted part structure allows the user to easily access the values, e.g. it makes it easy to do something like this, resulting in proper isolation of the formatted parts (not shown is decorating the parts separately):

```javascript
var message = // whatever the host node is for the string
for (let part of formattedMessage.parts) {
    var span = document.createElement('span');
    span.lang = part.lang;
    span.dir = part.dir;
    span.appendChild(document.createTextNode(part.value));
    message.appendChild(span);
}
```

That is done on portions of the message. The whole message also has a language (locale) and base paragraph direction.